Code/Tools/FBuild/FBuildCore/BFF/BFFBooleanExpParser.cpp

```diff
@@ -0,0 +1,959 @@
+static bool ParseStringComparisonExp( const Function * function, const OperandOf<String> & lhs, BFFTokenRange & iter, bool & expResult )
+{
+    ASSERT( lhs.IsValid() );
+
+    const BFFToken * opToken = iter.GetCurrent();
+    Operator op = Operator::OP_UNKNOWN;
+    if ( !ParseOperator( function, iter, op, &StringOperators ) )
+    {
+        return false;
+    }
+
+    // TODO: Support literal string arrays.
+    switch ( iter.GetCurrent()->GetType() )
+    {
+    case BFFTokenType::String:         // "Hello"
+    case BFFTokenType::Variable:       // .Var or ^Var
+    {
+        if ( ( op == Operator::OP_IN ) || ( op == Operator::OP_NOT_IN ) )
+        {
+            return ParseStringInArray( function, lhs, op, iter, expResult );
+        }
+
+        OperandOf<String> rhs(*iter.GetCurrent(), function);
+        if ( !rhs.EnsureValid() )
+        {
+            return false;
+        }
+        iter++;
+        return CompareStrings( function, lhs.GetValue(), rhs.GetValue(), opToken, op, expResult );
+    }
+    default:
+        Error::Error_1256_UnexpectedToken( iter.GetCurrent(), function );
+        return false;
+    }
+}
+
+/* ParseSubBooleanExp
+Forms:
+    bool-exp = (bool-exp)
+
+`iter` is expected pointing to the opening round bracket.
+------------------------------------------------------------------------------ */
+static bool ParseSubBooleanExp( const Function * function, BFFTokenRange & iter, bool & expResult )
+{
+    ASSERT( iter->GetType() == BFFTokenType::RoundBracket );
+
+    if ( iter->IsRoundBracket( ')' ) )
+    {
+        Error::Error_1061_MissingOpenBracket( iter.GetCurrent(), function );
+        return false;
+    }
+    ASSERT( iter->IsRoundBracket( '(' ) );
+    iter++;
+
+    if ( !ParseBooleanExp( function, iter, true, expResult ) )
+    {
+        return false;
+    }
+    if ( !iter->IsRoundBracket( ')' ) )
+    {
+        Error::Error_1002_MatchingClosingTokenNotFound( iter.GetCurrent(), function, ')' );
+        return false;
+    }
+    iter++;
+    return true;
+}
+
+/* ParseBooleanExp
+Forms:
+    bool-exp = true | false
+             | (bool-exp)
+             | bool-variable
+             | ! bool-exp
+             | bool-exp ==/!=/&&/|| bool-exp
+             | number-exp ==/!=/>/>=/</<= number-exp
+             | string-exp ==/!= string-exp
+             | string-exp in string-set
+             | string-exp not in string-set
+             | string-set in string-set
+             | string-set not in string-set
+    number-exp = number-literal | number-variable
+    string-exp = string-literal | string-variable
+    string-set = string-set-variable
+------------------------------------------------------------------------------ */
+static bool ParseBooleanExp( const Function * function, BFFTokenRange & iter, bool endsOnCloseBracket, bool & expResult )
+{
+    const BFFToken * const token = iter.GetCurrent();
+    ASSERT( token != nullptr ); // TODO: add a new error for "Unexpected end of expression".
```

Comment:
Generally the BFF parsing is written so this is impossible (there is always a sentinel end token).

Is it possible to get in here withe a nullptr token? If so, that might imply some bad logic at a higher level.

Reply:
I cannot remember the context for this line, it could be a misunderstanding on some random errors. 
I am going to remove it.

Reply:
Removed in 98ef95275848f20cea086e8c02cc3e13c8657003.
I will keep an eye on it in the future.